Add support for fake authentication

AMarcedone opened the pull request:
Adds support for fake authentication both to the client and the server (to be used for debug only). 

This allows the client to authenticate as user@domain.com by passing the additional parameter
`--fake-auth-userid=user@domain.com `
to the keytransparency-client binary (rather than having to get an actual oauth token). 
The feature should be enabled on the server side by starting the keytransparency-server binary with the ` --insecure-fake-auth ` flag.

AMarcedone commented:
Closes #603

gdbelvin commented on cmd/keytransparency-client/cmd/post.go:
```diff
@@ -20,6 +20,7 @@ import (
 	"log"
 	"time"
 
+	"github.com/google/keytransparency/core/authentication"
```
For consistency, please add a new line here.  We're organizing our imports as follows:

```go
import(
    "golang packages"
   
     "keytransparency packages"

     "other packages"
)
```

gdbelvin commented on cmd/keytransparency-client/cmd/root.go:
```diff
@@ -217,22 +219,32 @@ func dial(ktURL, caFile, clientSecretFile string, serviceKeyFile string) (*grpc.
+	// Add authentication information for the grpc.
+	fakeUserId := viper.GetString("fake-auth-userid")
+	if fakeUserId != "" { // Using fake authentication.
+		// Fake authentication information are added in cmd/keytransparency-client/cmd/post.go
```
It would be nice if the new fake auth implemented the [`credentials.PerRPCCredentials`](https://github.com/grpc/grpc-go/blob/master/credentials/credentials.go) interface so it could match the other methods of adding authenitcation to the connection.

AMarcedone replied:
The description of the interface is not very clear, but I looked at the existing implementations and hacked something similar :)

gdbelvin commented on cmd/keytransparency-client/cmd/root.go:
```diff
@@ -217,22 +219,32 @@ func dial(ktURL, caFile, clientSecretFile string, serviceKeyFile string) (*grpc.
+		// TODO(amarcedone) Consider having the GetClient method (or this method) initialize
+		// the context so that we can set the metadata parameters for the fake authentication
+		// here rather than in post.go.
+	} else {
```
This set of if/else's could be replaced with a `switch`, one case for each type of authentication.

gdbelvin commented on cmd/keytransparency-server/frontend.go:
```diff
@@ -140,9 +143,11 @@ func jsonLogger(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo
 
 	resp, err = handler(ctx, req)
 
+	/* TODO(amarcedone): logger not working. commenting out.
```
Just don't use the --verbose flag.  Does this produce a compile time error?

gdbelvin commented on cmd/keytransparency-server/frontend.go:
```diff
@@ -183,9 +188,17 @@ func main() {
-		glog.Exitf("Failed to create authentication library instance: %v", err)
+
+	var auth authentication.Authenticator
+	if *fakeAuth {
```
A switch here would be more appropriate.  I imagine we'll support multiple types of authentication in the future. 

Perhaps the flag could also be modified to be `auth-type=google` by default and `auth-type=fake` for testing.

AMarcedone replied:
Good idea =)

gdbelvin commented on core/authentication/fake.go:
```diff
@@ -31,24 +30,30 @@ func NewFake() *FakeAuth {
+}
+
+// Includes (fake) authentication information in the context to authenticate to the server as userID.
+func InsertFakeAuthIntoContext(ctx context.Context, userID string) context.Context {
```
Try making `FakeAuth` implement the `credentials.PerRPCCredentials` interface. 
That way, it could work seamlessly with the other GRPC authentication libraries.

AMarcedone commented:
I updated the PR with the requested changes.

Please note that in this commit https://github.com/google/keytransparency/pull/605/commits/3a12d1e58b26477515e3e747764d53cf071a3796
I removed the authentication context because (strangely) the test was not using it (even though it seemed to claim so as there was one test case which was tagged "Incorrect auth").

gdbelvin commented on core/authentication/fake.go:
```diff
@@ -29,26 +33,50 @@ func NewFake() *FakeAuth {
-	md := make(map[string][]string)
-	md["userid"] = []string{userID}
-	return metadata.NewOutgoingContext(context.TODO(), md)
+func GetFakeCredential(userID string) credentials.PerRPCCredentials{
```
For code clarity, would you mind moving this function below `FakeCredential`? 
We like to keep structs and their associated functions close together. Thanks!

gdbelvin commented on core/authentication/fake.go:
```diff
@@ -29,26 +33,50 @@ func NewFake() *FakeAuth {
 	return nil
 }
+
+const fakeCredentialType string = "FakeCredential"
```
Please move costs to the top of the file

AMarcedone commented:
Thanks a lot! I fixed an integration test which was failing and run gofmt on the files I touched :)

gdbelvin commented:
Thanks! LGTM and merged.

gdbelvin commented:
@jmillican This might be of interest to you
